### DISCUSSION TITLE
F601 `multi-value-repeated-key-literal` various issues

### DISCUSSION THREAD

**Discussion by dylwil3** [astral-sh/ruff#12918] [General]
This is an extension of #12772.

# Observed issues
[Playground Link](https://play.ruff.rs/e6b35fd7-dca9-421a-915a-fa3d34f04dd9)

- F601 should only be triggered for Expr's whose leaves are literals of some kind - the following two examples ought to fall under F602 instead.

```python
{(foo, bar): "a", (foo, bar): "b"}

{
    f"{foo}": "a",
    f"{foo}": "b",
}
```

- As in https://github.com/astral-sh/ruff/issues/12772 booleans and numbers that are equal should trigger this rule.

```python
{
    True: "a",
    1: "b",
    1.0: "c",
    1.0
    + 0.0j: "d",  # <-- this may be out of scope depending on the implementation of the fix
    2: "e",
    2.0: "f",
}
```

- (Maybe a feature request rather than a bug) It would be nice to extend F601 to situations like this, where expression ASTs have _literal_ leaves and internal nodes do not involve a function call

```python
{
    -1: "a",
    -1: "b",
}

{
    1 + 1: "a",
    1 + 1: "b",
}

# but _not_ to situations which require any kind of actual evaluation
{
    1 + 1: "a",
    2: "b",
}

# except maybe this, which appears above as well
# and which is arguably a casting/type promotion
{
    1: "a",
    1.0 + 0.0j: "b",
}
```

- There is also an issue with the suggested fix for values which are the "same", as there could be side effects, eg

```python
# Suggested fix is unsafe, since we can't know
# which value of iterator was desired
{
    1: next(it),
    1: next(it),
}
```

- Finally, a minor issue: the name of the rule does not seem quite right, since the linter complains if a key is repeated even if the same value appears.

# Potential Solution

I propose that the _logic_ of this check (independently of the implementation) should be as follows:

1. F601 should ignore keys which do not satisfy all of the following:
    - They are expressions that have only literals as leaves 
    - They are expressions where no node is a `Name` (that may be redundant since Name's are terminal?)
2. F601 should consider keys duplicated if the corresponding `ComparableExpr`'s are equal after further replacing all boolean and numeric literals with their complex-valued equivalent (for these purposes, I think "large" integers should be left as is).
3. F601 should not propose a fix if the "unique" value found has possible side effects - for example, we could just never propose a fix if the value found involves a function call.
4. Change the name of the rule to `repeated-key-literal` (and similarly for F602).

**Comment by dylwil3** [comment]
By the way, I'm happy to work on this I just wanna pin down the "assignment" first.

**Comment by MichaReiser** [comment]
I hope you don't mind if I convert this to a discussion.

**Comment by MichaReiser** [comment]
> F601 should only be triggered for Expr's whose leaves are literals of some kind - the following two examples ought to fall under F602 instead.

Can you tell me a bit more about the reasons why not:

```python
{(foo, bar): "a", (foo, bar): "b"}
```

Is it because the keys are tuples or because the tuples contain names?

```python
{
    f"{foo}": "a",
    f"{foo}": "b",
}
```

Is it because the f-strings have replacements? What about f-strings without replacements?

**Reply by dylwil3** [reply]
Sure - I may be misinterpreting the rule descriptions, but I thought that [F602](https://docs.astral.sh/ruff/rules/multi-value-repeated-key-variable/) was meant to catch repeated keys that involve names, whereas F601 was restricted to expressions involving literals alone. It just so happens that, in the implementation of  F601/F602, keys which are tuples or f-strings are always treated by F601:

https://github.com/astral-sh/ruff/blob/9b73532b112b5c2aa9f820bcfc1185d7af6ca02d/crates/ruff_linter/src/rules/pyflakes/rules/repeated_keys.rs#L156-L160

**Reply by MichaReiser** [reply]
Ah yeah, that seems incorrect. I think it should only handle constant-tuples and fstrings.

**Comment by MichaReiser** [comment]
>	````
>	{
>	    -1: "a",
>	    -1: "b",
>	}
>	
>	{
>	    1 + 1: "a",
>	    1 + 1: "b",
>	}
>	
>	# but _not_ to situations which require any kind of actual evaluation
>	{
>	    1 + 1: "a",
>	    2: "b",
>	}
>	
>	# except maybe this, which appears above as well
>	# and which is arguably a casting/type promotion
>	{
>	    1: "a",
>	    1.0 + 0.0j: "b",
>	}
>	````

IMO, explaining why the rule detects `1 + 1` and `1 + 1` but not `2` and `1 + 1` seems more difficult than just implementing the rule with some basic static-evaluation. Especially because the rule is about detecting identical literals (although it might need to be renamed because static evaluation goes slightly beyond).

**Reply by dylwil3** [reply]
I would be okay with that, in theory. It's not clear to me how difficult/expensive it is to perform this evaluation. Philosophically, this might also blur the line between "linter" and "type checker" (or even "compiler"), but that's also fine.

**Reply by MichaReiser** [reply]
Yeah. I also wonder if we already have something like that. @charliermarsh might now?

**Reply by MichaReiser** [reply]
This might be the closest that ruff has today (not very helpful in this context)

https://github.com/astral-sh/ruff/blob/aa0db338d935aa40fafaa05feff7343c42491d8e/crates/ruff_python_ast/src/helpers.rs#L1103-L1218

**Comment by MichaReiser** [comment]
> F601 should consider keys duplicated if the corresponding ComparableExpr's are equal after further replacing all boolean and numeric literals with their complex-valued equivalent (for these purposes, I think "large" integers should be left as is).

I overall agree with this except that allocating ad-hoc expressions is somewhat expensive. Considering that the rule only has to deal with a very limited set of expressions: would it make sense to create an ad-hoc comparison instead?

**Reply by dylwil3** [reply]
Sorry I was being a little unclear - this sentence was meant to give a criterion for an equivalence relation, but not specify how we implement checking equivalence. So I agree that two sorts of implementations would be:

1. Building a function `cast_nums(x: ComparableExp) -> ComparableExpr` and checking `cast_nums(x) = cast_nums(y)`.
2. Defining the new relation $\simeq$ directly as a function `is_iso: ((x,y):tuple[ComparableExpr,ComparableExpr]) -> bool`

In fact, the implementation of the rule involves the creation of a hash set, so instead of (2) one probably needs a custom hash function `h` which satisfies $x \simeq y \Rightarrow h(x) = h(y)$. That sounds like it would be more performant, I just don't know how to do that at the moment 😄 But I'd be happy to learn how, or learn that there is an easier way!

Regarding the issue of allocation: The implementation of the rule already allocates a `ComparableExpr` - if this "casting" function that replaces bools and  consumes it instead of borrowing it, would that resolve your concern? Sorry, I'm still developing an intuition for what's expensive and not here.

**Reply by MichaReiser** [reply]
Yeah I think a `HashMap`/`HashSet` makes sense here. You can get a custom hash function by creating a newtype-wrapper: A struct that wraps an inner type and you can implement `Eq`, `Hash`, etc. for it

**Reply by dylwil3** [reply]
Ok I'll give it a shot- sounds fun! And thanks for your help!
